# Class Imbalance Analysis

Analyze the distribution of the `failure_within_30_days` target to check whether or not the training dataset is imbalanced and examine how the failure examples are distributed across the drive models.

In [3]:
import pandas as pd

training_df = pd.read_csv("../data/processed/training_data.csv")

fail_count = len(training_df[training_df['failure_within_30_days'] == 1])
non_fail_count = len(training_df[training_df['failure_within_30_days'] == 0])
total_count = fail_count + non_fail_count

failure_percentage = (fail_count / total_count) * 100
non_failure_percentage = (non_fail_count / total_count) * 100


print(fail_count)
print(non_fail_count)

print(failure_percentage)
print(non_failure_percentage)

22786
79337
22.31230966579517
77.68769033420483


## Failures by Drive Model

Examine how the positive failure examples are distributed across different drive models to determine whether or not certain models make up a larger part of the failure data.

In [7]:
failure_rows = training_df[training_df['failure_within_30_days'] == 1]
drive_models = failure_rows['model']

print(drive_models.value_counts())

model
ST12000NM0008           2573
TOSHIBA MG08ACA16TA     2570
TOSHIBA MG07ACA14TA     2157
ST24000NM002H           1929
HGST HUH721212ALE604    1893
HGST HUH721212ALN604    1775
WDC WUH721816ALE6L4     1518
ST16000NM001G            980
WDC WUH722222ALE6L4      958
TOSHIBA MG10ACA20TE      836
ST12000NM001G            690
WDC WUH721816ALE6L0      646
ST8000NM0055             610
TOSHIBA MG08ACA16TEY     602
TOSHIBA MG08ACA16TE      535
ST14000NM0138            429
ST8000DM002              410
ST14000NM001G            356
ST12000NM0007            271
ST10000NM0086            227
HGST HUH721212ALE600     196
WDC WUH721414ALE6L4      159
TOSHIBA MG07ACA14TEY     125
WUH721816ALE6L4          124
TOSHIBA MG11ACA24TE       81
WDC WUH722626ALE6L4       58
ST12000NM000J             30
TOSHIBA MG09ACA16TE       16
ST8000NM000A              15
ST16000NM000J             11
ST14000NM000J              6
Name: count, dtype: int64


In [25]:
total_observations = (training_df['model']).value_counts()
positive_observations = drive_models.value_counts()

positive_rate_per_model = (positive_observations / total_observations) * 100
positive_rate_per_model = positive_rate_per_model.fillna(0).sort_values(ascending=False)

print(positive_rate_per_model)

model
WUH721816ALE6L4        54.625551
ST14000NM0138          45.541401
WDC WUH721816ALE6L0    41.251596
ST24000NM002H          39.946159
ST10000NM0086          35.248447
                         ...    
TOSHIBA MQ01ABF050M     0.000000
TOSHIBA MQ01ABF050      0.000000
WDC WD5000BPKT          0.000000
WDC WD5000LPVX          0.000000
WDC WDS250G2B0A         0.000000
Name: count, Length: 68, dtype: float64


## Conclusion

The training dataset contains 22,786 positive observations and 79,337 negative observations. About 22.3% of the observations are positive and 77.7% are negative, showing that the dataset has a class imbalance. The positive observations are also not distributed evenly across the drive models. Some models have much higher positive-label rates than others and some models have no positive observations in the training dataset. Since the majority class makes up about 77.7% of the dataset, the accuracy may be misleading when evaluating the machine learning model. A model that keeps predicting the majority class could have high accuracy while also still failing to identify drives that are approaching failure.